Array allocation sinking should split allocations into two, an Array allocation and a Butterfly allocation

kmiller68 opened the pull request:
#### c3b478c1983f5de5b1b216aa888ca5b47c738ac3
<pre>
Array allocation sinking should split allocations into two, an Array allocation and a Butterfly allocation
<a href="https://rdar.apple.com/159207754">rdar://159207754</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=298606">https://bugs.webkit.org/show_bug.cgi?id=298606</a>

Reviewed by Yusuke Suzuki.

This patch minorly rearchitects how we do Array allocation sinking in DFG. Previously we tried to model Arrays
in ObjectAllocationSinking as two allocations one where the actual Array was allocated and a &quot;Butterfly&quot;
at each `GetButterfly`. This meant that there was now a reverse data dependency between the GetButterfly and
the Array allocation Nodes. This was a little unintuitive but also meant that any control flow that would
merge two `GetButterfly`s would escape the Array.

This PR simplifies things by more directly representing the heap in ObjectAllocationSinking. There are now
two nodes that get sunk when sinking an Array: NewButterflyWithSize and NewArrayWithButterfly. All the
indexed properties and the Butterfly&apos;s location are stored on the NewArrayWithButterfly&apos;s Allocation.
The NewButterflyWithSize only contains the Array&apos;s length. If we ever wanted to extend Array allocation sinking
to support out of bounds stores we&apos;d have to teach `handleNode` how to deal with writes to multiple allocations.

Originally I had all the indexed properties and the length are stored on the LocalHeap of NewButterflyWithSize and
NewArrayWithButterfly&apos;s LocalHeap only contains the butterfly. But I realized that we crash while trying to materialize
a phantom array that contains itself or another newly materialized object that&apos;s in a cycle. This is because we have to
materialize the pointer as a PutByVal. We can&apos;t emit this PutByVal without the Array since we need to write barrier after
the store.

Overall, this change is perf neutral or maybe a slight progression on JetStream 3.

Canonical link: <a href="https://commits.webkit.org/300129@main">https://commits.webkit.org/300129@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4da3b01494a35b5899a62a5ca9372969534686af

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121381 "2 style errors") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41078 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31737 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127822 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73464 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cdc08f74-42f4-4aa0-8166-93372fcbcfce) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41780 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49657 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92201 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61348 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2b4d96af-ade7-4e9e-b742-a8103fa78311) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124333 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33363 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108760 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72876 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7ded78d2-1f7e-484e-b334-c6de5516e039) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32380 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26905 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71402 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/113511 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102858 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27079 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130655 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/119901 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48309 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36731 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100799 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48677 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104964 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100704 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/25537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46111 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24181 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45005 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19254 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48167 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53880 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150063 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47639 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38180 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50985 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49321 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGNodeType.h:
```diff
@@ -428,8 +428,9 @@ namespace JSC { namespace DFG {
     /* Support for allocation sinking. */\
     macro(PhantomNewButterflyWithSize, NodeResultStorage | NodeMustGenerate) \
-    macro(MaterializeNewButterflyWithSize, NodeResultStorage | NodeHasVarArgs) \
+    /* PhantomNewButterflyWithSize can materialize to NewButterflyWithSize as it's just a bookkeeper */ \
```
Maybe a bit wrong comment?

kmiller68 replied:
Changed to "PhantomNewButterflyWithSize can materialize to NewButterflyWithSize since it doesn't track any properties."

kmiller68 committed:
Array allocation sinking should split allocations into two, an Array allocation and a Butterfly allocation
rdar://159207754
https://bugs.webkit.org/show_bug.cgi?id=298606

Reviewed by Yusuke Suzuki.

This patch minorly rearchitects how we do Array allocation sinking in DFG. Previously we tried to model Arrays
in ObjectAllocationSinking as two allocations one where the actual Array was allocated and a "Butterfly"
at each `GetButterfly`. This meant that there was now a reverse data dependency between the GetButterfly and
the Array allocation Nodes. This was a little unintuitive but also meant that any control flow that would
merge two `GetButterfly`s would escape the Array.

This PR simplifies things by more directly representing the heap in ObjectAllocationSinking. There are now
two nodes that get sunk when sinking an Array: NewButterflyWithSize and NewArrayWithButterfly. All the
indexed properties and the Butterfly's location are stored on the NewArrayWithButterfly's Allocation.
The NewButterflyWithSize only contains the Array's length. If we ever wanted to extend Array allocation sinking
to support out of bounds stores we'd have to teach `handleNode` how to deal with writes to multiple allocations.

Originally I had all the indexed properties and the length are stored on the LocalHeap of NewButterflyWithSize and
NewArrayWithButterfly's LocalHeap only contains the butterfly. But I realized that we crash while trying to materialize
a phantom array that contains itself or another newly materialized object that's in a cycle. This is because we have to
materialize the pointer as a PutByVal. We can't emit this PutByVal without the Array since we need to write barrier after
the store.

Overall, this change is perf neutral or maybe a slight progression on JetStream 3.

Canonical link: https://commits.webkit.org/300129@main
